src/battle_controllers.c

```diff
@@ -3153,107 +3195,158 @@ void TrySetBattlerShadowSpriteCallback(u32 battler)
+        SetHealthboxSpriteVisible(gHealthboxSpriteIds[battler]);
 
-    if (side == B_SIDE_PLAYER)
-    {
-        CreateTask(Task_PlayerController_RestoreBgmAfterCry, 10);
-        HandleLowHpMusicChange(GetBattlerMon(battler), battler);
+        if (side == B_SIDE_OPPONENT)
+            CopyBattleSpriteInvisibility(battler);
+        gBattlerControllerFuncs[battler] = BtlController_HandleSwitchInShowSubstitute;
     }
+}
 
-    StartSpriteAnim(&gSprites[gBattlerSpriteIds[battler]], 0);
-    UpdateHealthboxAttribute(gHealthboxSpriteIds[battler], GetBattlerMon(battler), HEALTHBOX_ALL);
-    StartHealthboxSlideIn(battler);
-    SetHealthboxSpriteVisible(gHealthboxSpriteIds[battler]);
-
-    if (side == B_SIDE_OPPONENT)
+static void SwitchIn_CleanShinyAnimShowSubstitute(u32 battler)
+{
+    if (gSprites[gHealthboxSpriteIds[battler]].callback == SpriteCallbackDummy
+     && gBattleSpritesDataPtr->healthBoxesData[battler].finishedShinyMonAnim
+     && gSprites[gBattlerSpriteIds[battler]].callback == SpriteCallbackDummy)
+    {
         CopyBattleSpriteInvisibility(battler);
 
-    return TRUE;
+        // Reset shiny anim (even if it didn't occur)
+        gBattleSpritesDataPtr->healthBoxesData[battler].triedShinyMonAnim = FALSE;
+        gBattleSpritesDataPtr->healthBoxesData[battler].finishedShinyMonAnim = FALSE;
+        FreeSpriteTilesByTag(ANIM_TAG_GOLD_STARS);
+        FreeSpritePaletteByTag(ANIM_TAG_GOLD_STARS);
+
+        // Check if Substitute should be shown
+        if (gBattleSpritesDataPtr->battlerData[battler].behindSubstitute)
+            InitAndLaunchSpecialAnimation(battler, battler, battler, B_ANIM_MON_TO_SUBSTITUTE);
+
+        gBattlerControllerFuncs[battler] = BtlController_HandleSwitchInSoundAndEnd;
+    }
 }
 
-bool32 SwitchIn_TryShinyAnimUtil(u32 battler)
+void BtlController_HandleSwitchInTryShinyAnim(u32 battler)
 {
     if (!gBattleSpritesDataPtr->healthBoxesData[battler].ballAnimActive
      && !gBattleSpritesDataPtr->healthBoxesData[battler].triedShinyMonAnim)
         TryShinyAnimation(battler, GetBattlerMon(battler));
 
-    if (gSprites[gBattleControllerData[battler]].callback != SpriteCallbackDummy
-     || gBattleSpritesDataPtr->healthBoxesData[battler].ballAnimActive)
-        return FALSE;
-
-    DestroySprite(&gSprites[gBattleControllerData[battler]]);
+    if (gSprites[gBattleControllerData[battler]].callback == SpriteCallbackDummy
+     && !gBattleSpritesDataPtr->healthBoxesData[battler].ballAnimActive)
+    {
+        DestroySprite(&gSprites[gBattleControllerData[battler]]);
 
-    if (GetBattlerSide(battler) == B_SIDE_OPPONENT)
-        SetBattlerShadowSpriteCallback(battler, GetMonData(GetBattlerMon(battler), MON_DATA_SPECIES));
+        if (GetBattlerSide(battler) == B_SIDE_OPPONENT)
+            SetBattlerShadowSpriteCallback(battler, GetMonData(GetBattlerMon(battler), MON_DATA_SPECIES));
 
-    return TRUE;
+        if (gBattlerControllerEndFuncs[battler] == PlayerBufferExecCompleted)
```

Comment:
All good, but I don't like these checks. Maybe you could put them behind static inlines such as:
```C
static inline bool32 IsControllerPlayer(u32 battler)
{
    return (gBattlerControllerEndFuncs[battler] == PlayerBufferExecCompleted);
}
```
and that's for all the controllers checks, so it's more readable what we're actually checking

Reply:
Sounds good to me! :)